xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -178,78 +170,80 @@ type cdsBalancer struct {
+					cancelWatch := b.client.WatchCluster(update.clusterName, b.handleClusterUpdate)
+					b.logger.Infof("Watch started on resource name %v with xds-client %p", update.clusterName, b.client)
+					b.cancelWatch = func() {
+						cancelWatch()
+						b.logger.Infof("Watch cancelled on resource name %v with xds-client %p", update.clusterName, b.client)
+					}
+					b.clusterToWatch = update.clusterName
+				}
+			case *scUpdate:
 				if b.edsLB == nil {
-					b.logger.Errorf("Failed to create child policy of type %s, %v", edsName, err)
+					b.logger.Errorf("xds: received scUpdate {%+v} with no edsBalancer", update)
+					break
+				}
+				b.edsLB.UpdateSubConnState(update.subConn, update.state)
+			case *watchUpdate:
+				if err := update.err; err != nil {
+					b.logger.Warningf("Watch error from xds-client %p: %v", b.client, err)
+					b.handleErrorFromUpdate(err, false)
 					break
 				}
-				b.logger.Infof("Created child policy %p of type %s", b.edsLB, edsName)
-			}
-			lbCfg := &edsbalancer.EDSConfig{EDSServiceName: update.cds.ServiceName}
-			if update.cds.EnableLRS {
-				// An empty string here indicates that the edsBalancer
-				// should use the same xDS server for load reporting as
-				// it does for EDS requests/responses.
-				lbCfg.LrsLoadReportingServerName = new(string)
 
+				b.logger.Infof("Watch update from xds-client %p, content: %+v", b.client, update.cds)
+				// The first good update from the watch API leads to the
+				// instantiation of an edsBalancer. Further updates/errors are
+				// propagated to the existing edsBalancer.
+				if b.edsLB == nil {
+					var err error
+					b.edsLB, err = newEDSBalancer(b.cc, b.bOpts)
+					if b.edsLB == nil {
+						b.logger.Errorf("Failed to create child policy of type %s, %v", edsName, err)
+						break
+					}
+					b.logger.Infof("Created child policy %p of type %s", b.edsLB, edsName)
+				}
+				lbCfg := &edsbalancer.EDSConfig{EDSServiceName: update.cds.ServiceName}
+				if update.cds.EnableLRS {
+					// An empty string here indicates that the edsBalancer
+					// should use the same xDS server for load reporting as
+					// it does for EDS requests/responses.
+					lbCfg.LrsLoadReportingServerName = new(string)
+
+				}
+				ccState := balancer.ClientConnState{
+					ResolverState:  resolver.State{Attributes: attributes.New(xdsinternal.XDSClientID, b.client)},
+					BalancerConfig: lbCfg,
+				}
+				if err := b.edsLB.UpdateClientConnState(ccState); err != nil {
+					b.logger.Errorf("xds: edsBalancer.UpdateClientConnState(%+v) returned error: %v", ccState, err)
+				}
 			}
-			ccState := balancer.ClientConnState{
-				ResolverState:  resolver.State{Attributes: attributes.New(xdsinternal.XDSClientID, b.client)},
-				BalancerConfig: lbCfg,
-			}
-			if err := b.edsLB.UpdateClientConnState(ccState); err != nil {
-				b.logger.Errorf("xds: edsBalancer.UpdateClientConnState(%+v) returned error: %v", ccState, err)
-			}
-		case *closeUpdate:
+		case <-b.closed.Done():
```

Comment:
There's a 50/50 chance to keep processing the update after close. What makes things worse is, `b.updateCh` is a queue, so there can be several items in the queue, or added after close.

Check `b.closed` in `case u := <-b.updateCh.Get():`?

Reply:
We do check whether `closed` has fired before pushing stuff in to `b.updateCh`. So, items getting added to it after close doesn't seem to be possibility to me. And with this approach, we might close early even if there were stuff added to the `b.updateCh` before close was called, which I think is still fine. What do you think?


I could add another check for `b.closed` in `case u := <-b.updateCh.Get():`? But still not convinced that it is required ...

Reply:
Yeah, that makes sense. I missed that we won't push to the queue after close.